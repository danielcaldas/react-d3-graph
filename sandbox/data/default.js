module.exports = {
    links: [
        {
            source: 'Androsynth',
            target: 'Chenjesu'
        },
        {
            source: 'Androsynth',
            target: 'Ilwrath'
        },
        {
            source: 'Androsynth',
            target: 'Mycon'
        },
        {
            source: 'Androsynth',
            target: 'Spathi'
        },
        {
            source: 'Androsynth',
            target: 'Umgah'
        },
        {
            source: 'Androsynth',
            target: 'VUX'
        },
        {
            source: 'Chenjesu',
            target: 'Mycon'
        },
        {
            source: 'Chenjesu',
            target: 'Spathi'
        },
        {
            source: 'Chenjesu',
            target: 'Umgah'
        },
        {
            source: 'Chenjesu',
            target: 'VUX'
        },
        {
            source: 'Ilwrath',
            target: 'Spathi'
        },
        {
            source: 'Ilwrath',
            target: 'Umgah'
        },
        {
            source: 'Ilwrath',
            target: 'VUX'
        },
        {
            source: 'Mycon',
            target: 'Umgah'
        },
        {
            source: 'Mycon',
            target: 'VUX'
        },
        {
            source: 'Umgah',
            target: 'VUX'
        },
        {
            source: 'Androsynth',
            target: 'Guardian'
        },
        {
            source: 'Chenjesu',
            target: 'Broodhmome'
        },
        {
            source: 'Ilwrath',
            target: 'Avenger'
        },
        {
            source: 'Mycon',
            target: 'Podship'
        },
        {
            source: 'Spathi',
            target: 'Eluder'
        },
        {
            source: 'Umgah',
            target: 'Drone'
        },
        {
            source: 'VUX',
            target: 'Intruder'
        }
    ],
    nodes: [
        {
            id: 'Androsynth'
        },
        {
            id: 'Chenjesu'
        },
        {
            id: 'Ilwrath'
        },
        {
            id: 'Mycon'
        },
        {
            id: 'Spathi'
        },
        {
            id: 'Umgah'
        },
        {
            id: 'VUX'
        },
        {
            id: 'Guardian',
            symbolType: 'square'
        },
        {
            id: 'Broodhmome',
            symbolType: 'square'
        },
        {
            id: 'Avenger',
            symbolType: 'square'
        },
        {
            id: 'Podship',
            symbolType: 'square'
        },
        {
            id: 'Eluder',
            symbolType: 'square'
        },
        {
            id: 'Drone',
            symbolType: 'square'
        },
        {
            id: 'Intruder',
            symbolType: 'square'
        }
    ]
};
