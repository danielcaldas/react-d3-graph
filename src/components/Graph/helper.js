import React from 'react';

import * as d3 from 'd3';

import CONST from './const';

import Link from '../Link/';
import Node from '../Node/';

function buildGraph(nodes, nodeCallbacks, links, linkCallbacks, config, someNodeHighlighted) {
    let linksComponents = [];
    let nodesComponents = [];

    for (let node of Object.values(nodes)) {
        const props = buildNodeProps(node, config, nodeCallbacks, someNodeHighlighted);

        nodesComponents.push(<Node key={node.id} {...props} />);

        linksComponents = linksComponents.concat(buildNodeLinks(node, nodes, links, config, linkCallbacks, someNodeHighlighted));
    }

    return {
        nodes: nodesComponents,
        links: linksComponents
    };
}

function buildNodeProps(node, config, nodeCallbacks, someNodeHighlighted) {
    const opacity = someNodeHighlighted ? (node.highlighted ? config.node.opacity : config.highlightOpacity) : config.node.opacity;

    let fill = node.color || config.node.color;

    if (node.highlighted && config.node.highlightColor !== CONST.KEYWORDS.SAME) {
        fill = config.node.highlightColor;
    }

    return {
        className: CONST.NODE_CLASS_NAME,
        cursor: config.node.mouseCursor,
        cx: node && node.x || '0',
        cy: node && node.y || '0',
        fill,
        fontSize: node.highlighted ? config.node.highlightFontSize : config.node.fontSize,
        fontWeight: node.highlighted ? config.node.highlightFontWeight : config.node.fontWeight,
        id: node.id,
        label: node[config.node.labelProperty] || node.id,
        onClickNode: nodeCallbacks.onClickNode,
        onMouseOverNode: nodeCallbacks.onMouseOverNode,
        onMouseOut: nodeCallbacks.onMouseOut,
        opacity,
        renderLabel: config.node.renderLabel,
        size: node.size || config.node.size,
        stroke: node.highlighted ? config.node.highlightStrokeColor : config.node.strokeColor,
        strokeWidth: node.highlighted ? config.node.highlightStrokeWidth : config.node.strokeWidth,
        type: node.type || config.node.symbolType
    };
}

function buildNodeLinks(node, nodes, links, config, linkCallbacks, someNodeHighlighted) {
    let linksComponents = [];

    if (links[node.id]) {
        const x1 = node && node.x || '0';
        const y1 = node && node.y || '0';

        const adjacents = Object.keys(links[node.id]);
        const n = adjacents.length;

        for (let j=0; j < n; j++) {
            const target = adjacents[j];

            if (nodes[target]) {
                const key = `${node.id}${CONST.COORDS_SEPARATOR}${target}`;
                const props = buildLinkProps(node.id, target, x1, y1, nodes, links, config, linkCallbacks, someNodeHighlighted);

                linksComponents.push(<Link key={key} {...props} />);
            }
        }
    }

    return linksComponents;
}

function buildLinkProps(source, target, x1, y1, nodes, links, config, linkCallbacks, someNodeHighlighted) {
    const opacity = someNodeHighlighted ? (nodes[source].highlighted && nodes[target].highlighted) ? config.link.opacity : config.highlightOpacity : config.link.opacity;

    const stroke = (nodes[source].highlighted && nodes[target].highlighted) ?
                (config.link.highlightColor === CONST.KEYWORDS.SAME ? config.link.color : config.link.highlightColor)
                : config.link.color;

    const linkValue = links[source][target] || links[target][source];
    let strokeWidth = config.link.strokeWidth;

    if (config.link.semanticStrokeWidth) {
        strokeWidth += (linkValue * strokeWidth) / 10;
    }

    return {
        source,
        target,
        x1,
        y1,
        x2: nodes[target] && nodes[target].x || '0',
        y2: nodes[target] && nodes[target].y || '0',
        strokeWidth,
        stroke,
        className: CONST.LINK_CLASS_NAME,
        opacity,
        onClickLink: linkCallbacks.onClickLink
    };
}

function createForceSimulation(width, height) {
    const forceX = d3.forceX(width / 2).strength(CONST.FORCE_X);
    const forceY = d3.forceY(height / 2).strength(CONST.FORCE_Y);

    const simulation = d3.forceSimulation()
            .force('charge', d3.forceManyBody().strength(CONST.FORCE_IDEAL_STRENGTH))
            .force('x', forceX)
            .force('y', forceY);

    return simulation;
}

function initializeNodes(graphNodes) {
    let nodes = {};
    let indexMapping = {};
    let index = 0;

    graphNodes.forEach(n => {
        n['highlighted'] = false;
        if (!n.hasOwnProperty('x')) n['x'] = 0;
        if (!n.hasOwnProperty('y')) n['y'] = 0;

        nodes[n.id.toString()] = n;
        indexMapping[index] = n.id;

        index++;
    });

    return {
        nodes,
        indexMapping
    };
}

function initializeLinks(graphLinks) {
    let links = {};

    graphLinks.forEach(l => {
        if (!links[l.source]) {
            links[l.source] = {};
        }

        if (!links[l.target]) {
            links[l.target] = {};
        }

        // @TODO: If the graph is directed this should not happen
        links[l.source][l.target] = links[l.target][l.source] = l.value || 1;
    });

    return links;
}

export default {
    buildGraph,
    createForceSimulation,
    initializeLinks,
    initializeNodes
};
