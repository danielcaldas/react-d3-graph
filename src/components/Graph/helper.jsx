/**
 * @module Graph/helper
 * @description
 * Offers a series of methods that isolate logic of Graph component.
 */
/**
 * @typedef {Object} Link
 * @property {string} source - the node id of the source in the link.
 * @property {string} target - the node id of the target in the link.
 * @memberof Graph/helper
 */
/**
 * @typedef {Object} Node
 * @property {string} id - the id of the node.
 * @memberof Graph/helper
 */
import React from 'react';

import {
    forceX as d3ForceX,
    forceY as d3ForceY,
    forceSimulation as d3ForceSimulation,
    forceManyBody as d3ForceManyBody
} from 'd3-force';

import CONST from './const';
import DEFAULT_CONFIG from './config';
import ERRORS from '../../err';

import Link from '../link/';
import Node from '../node/';
import utils from '../../utils';

/**
 * Build some Link properties based on given parameters.
 * @param  {string} source - the id of the source node (from).
 * @param  {string} target - the id of the target node (to).
 * @param  {Object.<string, Object>} nodes - same as {@link #buildGraph|nodes in buildGraph}.
 * @param  {Object.<string, Object>} links - same as {@link #buildGraph|links in buildGraph}.
 * @param  {Object} config - same as {@link #buildGraph|config in buildGraph}.
 * @param  {Function[]} linkCallbacks - same as {@link #buildGraph|linkCallbacks in buildGraph}.
 * @param  {string} highlightedNode - same as {@link #buildGraph|highlightedNode in buildGraph}.
 * @param  {number} transform - value that indicates the amount of zoom transformation.
 * @returns {Object} returns an object that aggregates all props for creating respective Link component instance.
 * @memberof Graph/helper
 */
function _buildLinkProps(source, target, nodes, links, config, linkCallbacks, highlightedNode, transform) {
    const x1 = nodes[source] && nodes[source].x || 0;
    const y1 = nodes[source] && nodes[source].y || 0;
    const x2 = nodes[target] && nodes[target].x || 0;
    const y2 = nodes[target] && nodes[target].y || 0;

    let opacity = config.link.opacity;
    let mainNodeParticipates;

    switch (config.highlightDegree) {
        case 0:
            break;
        case 2:
            mainNodeParticipates = true;
            break;
        default: // 1st degree is the fallback behavior
            mainNodeParticipates = source === highlightedNode || target === highlightedNode;
            break;
    }

    if (highlightedNode) {
        opacity = (mainNodeParticipates
                && nodes[source].highlighted
                && nodes[target].highlighted) ? config.link.opacity : config.highlightOpacity;
    }

    let stroke = config.link.color;

    if (mainNodeParticipates && nodes[source].highlighted && nodes[target].highlighted) {
        stroke = config.link.highlightColor === CONST.KEYWORDS.SAME ? config.link.color
                                                                    : config.link.highlightColor;
    }

    const linkValue = links[source][target] || links[target][source] || 1;

    let strokeWidth = config.link.strokeWidth * (1 / transform);

    if (config.link.semanticStrokeWidth) {
        strokeWidth += (linkValue * strokeWidth) / 10;
    }

    return {
        source,
        target,
        x1,
        y1,
        x2,
        y2,
        strokeWidth,
        stroke,
        className: CONST.LINK_CLASS_NAME,
        opacity,
        onClickLink: linkCallbacks.onClickLink
    };
}

/**
 * Build Link components for a given node.
 * @param  {string} nodeId - the id of the node to whom Link components will be generated.
 * @param  {Object.<string, Object>} nodes - same as {@link #buildGraph|nodes in buildGraph}.
 * @param  {Object.<string, Object>} links - same as {@link #buildGraph|links in buildGraph}.
 * @param  {Object} config - same as {@link #buildGraph|config in buildGraph}.
 * @param  {Function[]} linkCallbacks - same as {@link #buildGraph|linkCallbacks in buildGraph}.
 * @param  {string} highlightedNode - same as {@link #buildGraph|highlightedNode in buildGraph}.
 * @param  {number} transform - value that indicates the amount of zoom transformation.
 * @returns {Object[]} returns the generated array of Link components.
 * @memberof Graph/helper
 */
function _buildNodeLinks(nodeId, nodes, links, config, linkCallbacks, highlightedNode, transform) {
    let linksComponents = [];

    if (links[nodeId]) {
        const adjacents = Object.keys(links[nodeId]);
        const n = adjacents.length;

        for (let j=0; j < n; j++) {
            const source = nodeId;
            const target = adjacents[j];

            if (nodes[target]) {
                const key = `${nodeId}${CONST.COORDS_SEPARATOR}${target}`;
                const props = _buildLinkProps(
                    source,
                    target,
                    nodes,
                    links,
                    config,
                    linkCallbacks,
                    highlightedNode,
                    transform
                );

                linksComponents.push(<Link key={key} {...props} />);
            }
        }
    }

    return linksComponents;
}

/**
 * Get the correct node opacity in order to properly make decisions based on context such as currently highlited node.
 * @param  {Object} node - the node object for whom we will generate properties.
 * @param  {string} highlightedNode - same as {@link #buildGraph|highlightedNode in buildGraph}.
 * @param  {Object} config - same as {@link #buildGraph|config in buildGraph}.
 * @returns {number} the opacity value for the given node.
 * @memberof Graph/helper
 */
function _getNodeOpacity(node, highlightedNode, config) {
    let opacity;

    if (highlightedNode && config.highlightDegree === 0) {
        opacity = (node.id === highlightedNode && node.highlighted) ? config.node.opacity : config.highlightOpacity;
    } else if (highlightedNode) {
        opacity = node.highlighted ? config.node.opacity : config.highlightOpacity;
    } else {
        opacity = config.node.opacity;
    }

    return opacity;
}

/**
 * Build some Node properties based on given parameters.
 * @param  {Object} node - the node object for whom we will generate properties.
 * @param  {Object} config - same as {@link #buildGraph|config in buildGraph}.
 * @param  {Function[]} nodeCallbacks - same as {@link #buildGraph|nodeCallbacks in buildGraph}.
 * @param  {string} highlightedNode - same as {@link #buildGraph|highlightedNode in buildGraph}.
 * @param  {number} transform - value that indicates the amount of zoom transformation.
 * @returns {Object} returns object that contain Link props ready to be feeded to the Link component.
 * @memberof Graph/helper
 */
function _buildNodeProps(node, config, nodeCallbacks, highlightedNode, transform) {
    const opacity = _getNodeOpacity(node, highlightedNode, config);

    let fill = node.color || config.node.color;

    if (node.highlighted && config.node.highlightColor !== CONST.KEYWORDS.SAME) {
        fill = config.node.highlightColor;
    }

    let stroke = config.node.strokeColor;

    if (node.highlighted && config.node.highlightStrokeColor !== CONST.KEYWORDS.SAME) {
        stroke = config.node.highlightStrokeColor;
    }

    const t = 1 / transform;
    const nodeSize = node.size || config.node.size;
    const fontSize = node.highlighted ? config.node.highlightFontSize : config.node.fontSize;
    const dx = (fontSize * t) + (nodeSize / 100) + 1.5;
    const strokeWidth = node.highlighted ? config.node.highlightStrokeWidth : config.node.strokeWidth;

    return {
        className: CONST.NODE_CLASS_NAME,
        cursor: config.node.mouseCursor,
        cx: node && node.x || '0',
        cy: node && node.y || '0',
        fill,
        fontSize: fontSize * t,
        dx,
        fontWeight: node.highlighted ? config.node.highlightFontWeight : config.node.fontWeight,
        id: node.id,
        label: node[config.node.labelProperty] || node.id,
        onClickNode: nodeCallbacks.onClickNode,
        onMouseOverNode: nodeCallbacks.onMouseOverNode,
        onMouseOut: nodeCallbacks.onMouseOut,
        opacity,
        renderLabel: config.node.renderLabel,
        size: nodeSize * t,
        stroke,
        strokeWidth: strokeWidth * t,
        type: node.symbolType || config.node.symbolType
    };
}

/**
 * Method that actually is exported an consumed by Graph component in order to build all Nodes and Link
 * components.
 * @param  {Object.<string, Object>} nodes - an object containing all nodes mapped by their id.
 * @param  {Function[]} nodeCallbacks - array of callbacks for used defined event handler for node interactions.
 * @param  {Object.<string, Object>} links - an object containing a matrix of connections of the graph, for each nodeId,
 * there is an Object that maps adjacent nodes ids (string) and their values (number).
 * ```javascript
 *  // links example
 *  {
 *     "Androsynth": {
 *         "Chenjesu": 1,
 *         "Ilwrath": 1,
 *         "Mycon": 1,
 *         "Spathi": 1,
 *         "Umgah": 1,
 *         "VUX": 1,
 *         "Guardian": 1
 *     },
 *     "Chenjesu": {
 *         "Androsynth": 1,
 *         "Mycon": 1,
 *         "Spathi": 1,
 *         "Umgah": 1,
 *         "VUX": 1,
 *         "Broodhmome": 1
 *     },
 *     ...
 *  }
 * ```
 * @param  {Function[]} linkCallbacks - array of callbacks for used defined event handler for link interactions.
 * @param  {Object} config - an object containg rd3g consumer defined configurations {@link #config config} for the graph.
 * @param  {string} highlightedNode - this value contains a string that represents the some currently highlighted node.
 * @param  {number} transform - value that indicates the amount of zoom transformation.
 * @returns {Object} returns an object containg the generated nodes and links that form the graph. The result is
 * returned in a way that can be consumed by es6 **destructuring assignment**.
 * @memberof Graph/helper
 */
function buildGraph(nodes, nodeCallbacks, links, linkCallbacks, config, highlightedNode, transform) {
    let linksComponents = [];
    let nodesComponents = [];

    for (let i = 0, keys = Object.keys(nodes), n = keys.length; i < n; i++) {
        const nodeId = keys[i];

        const props = _buildNodeProps(nodes[nodeId], config, nodeCallbacks, highlightedNode, transform);

        nodesComponents.push(<Node key={nodeId} {...props} />);

        linksComponents = linksComponents.concat(
            _buildNodeLinks(nodeId, nodes, links, config, linkCallbacks, highlightedNode, transform)
        );
    }

    return {
        nodes: nodesComponents,
        links: linksComponents
    };
}

/**
 * Create d3 forceSimulation to be applied on the graph.<br/>
 * <a href="https://github.com/d3/d3-force#forceSimulation" target="_blank">https://github.com/d3/d3-force#forceSimulation</a><br/>
 * <a href="https://github.com/d3/d3-force#simulation_force" target="_blank">https://github.com/d3/d3-force#simulation_force</a><br/>
 * @param  {number} width - the width of the container area of the graph.
 * @param  {number} height - the height of the container area of the graph.
 * @returns {Object} returns the simulation instance to be consumed.
 * @memberof Graph/helper
 */
function createForceSimulation(width, height) {
    const frx = d3ForceX(width / 2).strength(CONST.FORCE_X);
    const fry = d3ForceY(height / 2).strength(CONST.FORCE_Y);

    return d3ForceSimulation()
            .force('charge', d3ForceManyBody().strength(CONST.FORCE_IDEAL_STRENGTH))
            .force('x', frx)
            .force('y', fry);
}

/**
 * Incapsulates common procedures to initialize graph.
 * @param {Object} props - Graph component props, object that holds data, id and config.
 * @param {Object} props.data - Data object holds links (array of **Link**) and nodes (array of **Node**).
 * @param {string} props.id - the graph id.
 * @param {Object} props.config - same as {@link #buildGraph|config in buildGraph}.
 * @param {Object} state - Graph component current state (same format as returned object on this function).
 * @returns a fully (re)initialized graph state object.
 * @memberof Graph/helper
 */
function initializeGraphState({data, id, config}, state) {
    let graph;

    validateGraphData(data);

    if (state && state.nodes && state.links && state.nodeIndexMapping) {
        // absorve existent positining
        graph = {
            nodes: data.nodes.map(n => Object.assign({}, n, state.nodes[n.id])),
            links: {}
        };
    } else {
        graph = {
            nodes: data.nodes.map(n => Object.assign({}, n)),
            links: {}
        };
    }

    graph.links = data.links.map(l => Object.assign({}, l));

    let newConfig = Object.assign({}, utils.merge(DEFAULT_CONFIG, config || {}));
    let {nodes, nodeIndexMapping} = initializeNodes(graph.nodes);
    let links = initializeLinks(graph.links); // Matrix of graph connections
    const {nodes: d3Nodes, links: d3Links} = graph;
    const formatedId = id.replace(/ /g, '_');
    const simulation = createForceSimulation(newConfig.width, newConfig.height);

    return {
        id: formatedId,
        config: newConfig,
        nodeIndexMapping,
        links,
        d3Links,
        nodes,
        d3Nodes,
        highlightedNode: '',
        simulation,
        newGraphElements: false,
        configUpdated: false,
        transform: 1
    };
}

/**
 * Receives a matrix of the graph with the links source and target as concrete node instances and it transforms it
 * in a lightweight matrix containing only links with source and target being strings representative of some node id
 * and the respective link value (if non existant will default to 1).
 * @param  {Object[]} graphLinks - an array of all graph links but all the links contain the source and target nodes
 * objects.
 * @returns {Object.<string, Object>} an object containing a matrix of connections of the graph, for each nodeId,
 * there is an object that maps adjacent nodes ids (string) and their values (number).
 * @memberof Graph/helper
 */
function initializeLinks(graphLinks) {
    return graphLinks.reduce((links, l) => {
        const source = l.source.id || l.source;
        const target = l.target.id || l.target;

        if (!links[source]) {
            links[source] = {};
        }

        if (!links[target]) {
            links[target] = {};
        }

        // @TODO: If the graph is directed this should be adapted
        links[source][target] = links[target][source] = l.value || 1;

        return links;
    }, {});
}

/**
 * Method that initialize graph nodes provided by rd3g consumer and adds additional default mandatory properties
 * that are optional for the user. Also it generates an index mapping, this maps nodes ids the their index in the array
 * of nodes. This is needed because d3 callbacks such as node click and link click return the index of the node.
 * @param  {Object[]} graphNodes - the array of nodes provided by the rd3g consumer.
 * @returns {Object} returns the nodes ready to be used within rd3g with additional properties such as x, y
 * and highlighted values. Returns also the index mapping object of type Object.<number, string>.
 * @memberof Graph/helper
 */
function initializeNodes(graphNodes) {
    let nodes = {};
    let nodeIndexMapping = {};

    const n = graphNodes.length;

    for (let i=0; i < n; i++) {
        const node = graphNodes[i];

        node.highlighted = false;

        if (!node.hasOwnProperty('x')) { node['x'] = 0; }
        if (!node.hasOwnProperty('y')) { node['y'] = 0; }

        nodes[node.id.toString()] = node;
        nodeIndexMapping[i] = node.id;
    }

    return { nodes, nodeIndexMapping };
}

/**
 * Some integraty validations on links and nodes structure. If some validation fails the function will
 * throw an error.
 * @param  {Object} data - Same as {@link #initializeGraphState|data in initializeGraphState}.
 * @memberof Graph/helper
 */
function validateGraphData(data) {
    data.links.forEach(l => {
        if (!data.nodes.find(n => n.id === l.source)) {
            utils.throwErr(this.constructor.name, `${ERRORS.INVALID_LINKS} - ${l.source} is not a valid node id`);
        }
        if (!data.nodes.find(n => n.id === l.target)) {
            utils.throwErr(this.constructor.name, `${ERRORS.INVALID_LINKS} - ${l.target} is not a valid node id`);
        }
    });
}

export default {
    buildGraph,
    createForceSimulation,
    initializeGraphState,
    initializeLinks,
    initializeNodes
};
