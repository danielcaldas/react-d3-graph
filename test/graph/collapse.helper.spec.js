import * as collapseHelper from "../../src/components/graph/collapse.helper";
import * as graphData from "../../sandbox/data/default";
import { initializeGraphState } from "../../src/components/graph/graph.helper";

describe("Collapse Helper", () => {
    let directedState;
    let defaultState;

    beforeEach(() => {
        directedState = initializeGraphState({ data: graphData, id: "id", config: { directed: true } });
        defaultState = initializeGraphState({ data: graphData, id: "id" });
    });

    describe("#computeNodeDegree", () => {
        describe("when graph is directed", () => {
            test("should calculate correct node degree for 'Androsynth'", () => {
                const nodeId = "Androsynth";

                expect(collapseHelper.computeNodeDegree(nodeId, directedState.links)).toEqual({
                    inDegree: 0,
                    outDegree: 7,
                });
            });

            test("should calculate correct node degree for 'Podship'", () => {
                const nodeId = "Podship";

                expect(collapseHelper.computeNodeDegree(nodeId, directedState.links)).toEqual({
                    inDegree: 1,
                    outDegree: 0,
                });
            });

            test("should calculate correct node degree for 'Mycon'", () => {
                const nodeId = "Mycon";

                expect(collapseHelper.computeNodeDegree(nodeId, directedState.links)).toEqual({
                    inDegree: 2,
                    outDegree: 3,
                });
            });
        });

        describe("when graph is not directed", () => {
            test("should calculate correct node degree for 'Androsynth'", () => {
                const nodeId = "Androsynth";

                expect(collapseHelper.computeNodeDegree(nodeId, defaultState.links)).toEqual({
                    inDegree: 7,
                    outDegree: 7,
                });
            });

            test("should calculate correct node degree for 'Podship'", () => {
                const nodeId = "Podship";

                expect(collapseHelper.computeNodeDegree(nodeId, defaultState.links)).toEqual({
                    inDegree: 1,
                    outDegree: 1,
                });
            });

            test("should calculate correct node degree for 'Mycon'", () => {
                const nodeId = "Mycon";

                expect(collapseHelper.computeNodeDegree(nodeId, defaultState.links)).toEqual({
                    inDegree: 5,
                    outDegree: 5,
                });
            });
        });
    });

    describe("#getTargetLeafConnections", () => {
        let directed;

        describe("when graph is directed", () => {
            beforeAll(() => {
                directed = true;
            });
            test("should return expected leaf connections for node 'Mycon'", () => {
                const nodeId = "Mycon";

                expect(collapseHelper.getTargetLeafConnections(nodeId, directedState.links, { directed })).toEqual([
                    { source: "Mycon", target: "Podship" },
                ]);
            });

            test("should return expected leaf connections for node 'VUX'", () => {
                const nodeId = "VUX";

                expect(collapseHelper.getTargetLeafConnections(nodeId, directedState.links, { directed })).toEqual([
                    { source: "VUX", target: "Intruder" },
                ]);
            });

            test("should return expected leaf connections for node 'Eluder'", () => {
                const nodeId = "Eluder";

                expect(collapseHelper.getTargetLeafConnections(nodeId, directedState.links, { directed })).toEqual([]);
            });
        });

        describe("when graph is not directed", () => {
            beforeAll(() => {
                directed = false;
            });

            test("should calculate correct node degree for 'Androsynth'", () => {
                const nodeId = "Mycon";

                expect(collapseHelper.getTargetLeafConnections(nodeId, defaultState.links, { directed })).toEqual([
                    { source: "Mycon", target: "Podship" },
                ]);
            });

            test("should return expected leaf connections for node 'VUX'", () => {
                const nodeId = "VUX";

                expect(collapseHelper.getTargetLeafConnections(nodeId, defaultState.links, { directed })).toEqual([
                    { source: "VUX", target: "Intruder" },
                ]);
            });

            test("should return expected leaf connections for node 'Eluder'", () => {
                const nodeId = "Eluder";

                expect(collapseHelper.getTargetLeafConnections(nodeId, defaultState.links, { directed })).toEqual([]);
            });
        });
    });

    describe("#toggleLinksConnections", () => {
        test("should properly set isHidden value for given links and linksMatrix", () => {
            const d3Links = directedState.d3Links;
            const connectionsMatrix = directedState.links;

            connectionsMatrix["Androsynth"]["Ilwrath"] = 0; // d3Links[1] isHidden: true
            delete connectionsMatrix["Androsynth"]["Chenjesu"]; // d3Links[0] isHidden: true

            const updatedD3Links = collapseHelper.toggleLinksConnections(d3Links, connectionsMatrix);

            expect(updatedD3Links[1].isHidden).toEqual(true);
            expect(updatedD3Links[0].isHidden).toEqual(true);

            expect(updatedD3Links).toMatchSnapshot();
        });
    });

    describe("#toggleLinksMatrixConnections", () => {
        describe("when graph is directed", () => {
            test("should properly toggle passed connections in links matrix", () => {
                const connections = [
                    { source: "Androsynth", target: "Chenjesu" },
                    { source: "Androsynth", target: "Ilwrath" },
                ];

                expect(
                    collapseHelper.toggleLinksMatrixConnections(directedState.links, connections, { directed: true })
                ).toMatchSnapshot();
            });
        });

        describe("when graph is not directed", () => {
            test("should properly toggle passed connections in links matrix", () => {
                const connections = [
                    { source: "Androsynth", target: "Chenjesu" },
                    { source: "Androsynth", target: "Ilwrath" },
                ];

                expect(
                    collapseHelper.toggleLinksMatrixConnections(defaultState.links, connections, { directed: false })
                ).toMatchSnapshot();
            });
        });
    });
});
