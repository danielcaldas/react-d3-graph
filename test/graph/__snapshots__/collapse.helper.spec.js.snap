// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collapse Helper #toggleLinksConnections should properly set isHidden value for given links and linksMatrix 1`] = `
Array [
  Object {
    "isHidden": true,
    "label": "connects to Chenjesu",
    "source": "Androsynth",
    "target": "Chenjesu",
  },
  Object {
    "isHidden": true,
    "label": "connects to Ilwrath",
    "source": "Androsynth",
    "target": "Ilwrath",
  },
  Object {
    "isHidden": false,
    "label": "connects to Mycon",
    "source": "Androsynth",
    "target": "Mycon",
  },
  Object {
    "isHidden": false,
    "label": "connects to Spathi",
    "source": "Androsynth",
    "target": "Spathi",
  },
  Object {
    "isHidden": false,
    "source": "Androsynth",
    "target": "Umgah",
  },
  Object {
    "isHidden": false,
    "source": "Androsynth",
    "target": "VUX",
  },
  Object {
    "isHidden": false,
    "source": "Chenjesu",
    "target": "Mycon",
  },
  Object {
    "isHidden": false,
    "source": "Chenjesu",
    "target": "Spathi",
  },
  Object {
    "isHidden": false,
    "source": "Chenjesu",
    "target": "Umgah",
  },
  Object {
    "isHidden": false,
    "source": "Chenjesu",
    "target": "VUX",
  },
  Object {
    "isHidden": false,
    "source": "Ilwrath",
    "target": "Spathi",
  },
  Object {
    "isHidden": false,
    "source": "Ilwrath",
    "target": "Umgah",
  },
  Object {
    "isHidden": false,
    "source": "Ilwrath",
    "target": "VUX",
  },
  Object {
    "isHidden": false,
    "source": "Mycon",
    "target": "Umgah",
  },
  Object {
    "isHidden": false,
    "source": "Mycon",
    "target": "VUX",
  },
  Object {
    "isHidden": false,
    "source": "Umgah",
    "target": "VUX",
  },
  Object {
    "isHidden": false,
    "source": "Androsynth",
    "target": "Guardian",
  },
  Object {
    "isHidden": false,
    "source": "Chenjesu",
    "target": "Broodhmome",
  },
  Object {
    "isHidden": false,
    "source": "Ilwrath",
    "target": "Avenger",
  },
  Object {
    "isHidden": false,
    "source": "Mycon",
    "target": "Podship",
  },
  Object {
    "isHidden": false,
    "source": "Spathi",
    "target": "Eluder",
  },
  Object {
    "isHidden": false,
    "source": "Umgah",
    "target": "Drone",
  },
  Object {
    "isHidden": false,
    "source": "VUX",
    "target": "Intruder",
  },
]
`;

exports[`Collapse Helper #toggleLinksMatrixConnections when graph is directed should properly toggle passed connections in links matrix 1`] = `
Object {
  "Androsynth": Object {
    "Chenjesu": 0,
    "Guardian": 1,
    "Ilwrath": 0,
    "Mycon": 1,
    "Spathi": 1,
    "Umgah": 1,
    "VUX": 1,
  },
  "Avenger": Object {},
  "Broodhmome": Object {},
  "Chenjesu": Object {
    "Broodhmome": 1,
    "Mycon": 1,
    "Spathi": 1,
    "Umgah": 1,
    "VUX": 1,
  },
  "Drone": Object {},
  "Eluder": Object {},
  "Guardian": Object {},
  "Ilwrath": Object {
    "Avenger": 1,
    "Spathi": 1,
    "Umgah": 1,
    "VUX": 1,
  },
  "Intruder": Object {},
  "Mycon": Object {
    "Podship": 1,
    "Umgah": 1,
    "VUX": 1,
  },
  "Podship": Object {},
  "Spathi": Object {
    "Eluder": 1,
  },
  "Umgah": Object {
    "Drone": 1,
    "VUX": 1,
  },
  "VUX": Object {
    "Intruder": 1,
  },
}
`;

exports[`Collapse Helper #toggleLinksMatrixConnections when graph is not directed should properly toggle passed connections in links matrix 1`] = `
Object {
  "Androsynth": Object {
    "Chenjesu": 0,
    "Guardian": 1,
    "Ilwrath": 0,
    "Mycon": 1,
    "Spathi": 1,
    "Umgah": 1,
    "VUX": 1,
  },
  "Avenger": Object {
    "Ilwrath": 1,
  },
  "Broodhmome": Object {
    "Chenjesu": 1,
  },
  "Chenjesu": Object {
    "Androsynth": 0,
    "Broodhmome": 1,
    "Mycon": 1,
    "Spathi": 1,
    "Umgah": 1,
    "VUX": 1,
  },
  "Drone": Object {
    "Umgah": 1,
  },
  "Eluder": Object {
    "Spathi": 1,
  },
  "Guardian": Object {
    "Androsynth": 1,
  },
  "Ilwrath": Object {
    "Androsynth": 0,
    "Avenger": 1,
    "Spathi": 1,
    "Umgah": 1,
    "VUX": 1,
  },
  "Intruder": Object {
    "VUX": 1,
  },
  "Mycon": Object {
    "Androsynth": 1,
    "Chenjesu": 1,
    "Podship": 1,
    "Umgah": 1,
    "VUX": 1,
  },
  "Podship": Object {
    "Mycon": 1,
  },
  "Spathi": Object {
    "Androsynth": 1,
    "Chenjesu": 1,
    "Eluder": 1,
    "Ilwrath": 1,
  },
  "Umgah": Object {
    "Androsynth": 1,
    "Chenjesu": 1,
    "Drone": 1,
    "Ilwrath": 1,
    "Mycon": 1,
    "VUX": 1,
  },
  "VUX": Object {
    "Androsynth": 1,
    "Chenjesu": 1,
    "Ilwrath": 1,
    "Intruder": 1,
    "Mycon": 1,
    "Umgah": 1,
  },
}
`;
